File: mac/Sources/SoundboyCore/MediaFiles.swift
import Foundation

public enum MediaFiles {
    /// Formats Core Audio decodes natively (no OGG/WMA on macOS).
    public static let audioExtensions: Set<String> = [
        "mp3", "m4a", "m4b", "aac", "mp4", "wav", "aif", "aiff", "aifc", "caf", "flac",
    ]
    public static let playlistExtensions: Set<String> = ["m3u", "m3u8", "pls"]

    public static func isAudio(_ url: URL) -> Bool { audioExtensions.contains(url.pathExtension.lowercased()) }
    public static func isPlaylist(_ url: URL) -> Bool { playlistExtensions.contains(url.pathExtension.lowercased()) }

    /// Turns files, folders (recursive) and playlists into tracks.
    public static func expand(_ urls: [URL]) -> [Track] {
        var result: [Track] = []
        let fm = FileManager.default
        for url in urls {
            var isDir: ObjCBool = false
            guard fm.fileExists(atPath: url.path, isDirectory: &isDir) else { continue }
            if isDir.boolValue {
                var files: [URL] = []
                if let e = fm.enumerator(at: url, includingPropertiesForKeys: [.isRegularFileKey], options: [.skipsHiddenFiles]) {
                    for case let f as URL in e where isAudio(f) { files.append(f) }
                }
                files.sort { $0.path.localizedStandardCompare($1.path) == .orderedAscending }
                result += files.map(Track.init)
            } else if isPlaylist(url) {
                result += (try? PlaylistIO.load(url)) ?? []
            } else if isAudio(url) {
                result.append(Track(url: url))
            }
        }
        return result
    }
}

public enum PlaylistIO {
    public static func load(_ url: URL) throws -> [Track] {
        let data = try Data(contentsOf: url)
        let text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
        let lines = text.replacingOccurrences(of: "\u{FEFF}", with: "").components(separatedBy: .newlines)
        let dir = url.deletingLastPathComponent()
        return url.pathExtension.lowercased() == "pls" ? loadPls(lines, dir) : loadM3u(lines, dir)
    }

    static func loadM3u(_ lines: [String], _ dir: URL) -> [Track] {
        var list: [Track] = []
        var title: String?
        var duration: TimeInterval?
        for raw in lines {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.isEmpty { continue }
            if line.uppercased().hasPrefix("#EXTINF:") {
                let body = String(line.dropFirst(8))
                if let comma = body.firstIndex(of: ",") {
                    let secs = body[..<comma].split(separator: " ").first.map(String.init) ?? ""
                    if let s = Double(secs), s > 0 { duration = s }
                    title = String(body[body.index(after: comma)...]).trimmingCharacters(in: .whitespaces)
                }
                continue
            }
            if line.hasPrefix("#") { continue }
            if let t = makeTrack(line, dir, title, duration) { list.append(t) }
            title = nil
            duration = nil
        }
        return list
    }

    static func loadPls(_ lines: [String], _ dir: URL) -> [Track] {
        var files: [Int: String] = [:], titles: [Int: String] = [:], lengths: [Int: Double] = [:]
        for raw in lines {
            guard let eq = raw.firstIndex(of: "=") else { continue }
            let key = raw[..<eq].trimmingCharacters(in: .whitespaces).lowercased()
            let val = raw[raw.index(after: eq)...].trimmingCharacters(in: .whitespaces)
            for (prefix, apply) in [("file", { (n: Int) in files[n] = val }),
                                    ("title", { (n: Int) in titles[n] = val }),
                                    ("length", { (n: Int) in lengths[n] = Double(val) })] {
                if key.hasPrefix(prefix), let n = Int(key.dropFirst(prefix.count)) { apply(n) }
            }
        }
        return files.keys.sorted().compactMap { n in
            makeTrack(files[n]!, dir, titles[n], lengths[n].flatMap { $0 > 0 ? $0 : nil })
        }
    }

    static func makeTrack(_ entry: String, _ dir: URL, _ title: String?, _ duration: TimeInterval?) -> Track? {
        let url: URL
        if entry.lowercased().hasPrefix("file:") {
            guard let u = URL(string: entry) else { return nil }
            url = u
        } else if entry.hasPrefix("/") {
            url = URL(fileURLWithPath: entry)
        } else if entry.contains("://") {
            return nil // streams are not supported on macOS yet
        } else {
            url = dir.appendingPathComponent(entry.replacingOccurrences(of: "\\", with: "/")).standardizedFileURL
        }
        let t = Track(url: url)
        t.duration = duration
        if let title, !title.isEmpty { t.title = title }
        return t
    }

    public static func save(_ tracks: [Track], to url: URL) throws {
        var out = ""
        if url.pathExtension.lowercased() == "pls" {
            out += "[playlist]\n"
            for (i, t) in tracks.enumerated() {
                out += "File\(i + 1)=\(t.url.path)\nTitle\(i + 1)=\(t.displayTitle)\nLength\(i + 1)=\(Int(t.duration ?? -1))\n"
            }
            out += "NumberOfEntries=\(tracks.count)\nVersion=2\n"
        } else {
            out += "#EXTM3U\n"
            for t in tracks {
                out += "#EXTINF:\(t.duration.map { Int($0) } ?? -1),\(t.displayTitle)\n\(t.url.path)\n"
            }
        }
        try out.write(to: url, atomically: true, encoding: .utf8)
    }
}

File: SOUNDBOY/Model/MediaFiles.cs
using System.Globalization;
using System.Text;

namespace SOUNDBOY.Model;

public static class MediaFiles
{
    public static readonly string[] AudioExtensions =
    {
        ".mp3", ".mp2", ".wav", ".flac", ".ogg", ".oga", ".m4a", ".m4b", ".aac", ".mp4", ".wma", ".aif", ".aiff", ".ac3",
    };

    public static readonly string[] PlaylistExtensions = { ".m3u", ".m3u8", ".pls" };

    public static bool IsUrl(string s) =>
        s.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
        s.StartsWith("https://", StringComparison.OrdinalIgnoreCase);

    public static bool IsAudio(string path) =>
        AudioExtensions.Contains(Path.GetExtension(path), StringComparer.OrdinalIgnoreCase);

    public static bool IsPlaylist(string path) =>
        PlaylistExtensions.Contains(Path.GetExtension(path), StringComparer.OrdinalIgnoreCase);

    /// <summary>Turns files, folders (recursive), playlists and URLs into tracks.</summary>
    public static IEnumerable<Track> Expand(IEnumerable<string> paths)
    {
        foreach (var raw in paths)
        {
            var p = raw.Trim().Trim('"');
            if (p.Length == 0) continue;
            if (IsUrl(p)) { yield return new Track(p); continue; }

            if (Directory.Exists(p))
            {
                List<string> files;
                try
                {
                    files = Directory
                        .EnumerateFiles(p, "*", new EnumerationOptions { RecurseSubdirectories = true, IgnoreInaccessible = true })
                        .Where(IsAudio)
                        .OrderBy(f => f, StringComparer.CurrentCultureIgnoreCase)
                        .ToList();
                }
                catch { continue; }
                foreach (var f in files) yield return new Track(f);
                continue;
            }

            if (IsPlaylist(p) && File.Exists(p))
            {
                List<Track> list;
                try { list = PlaylistIO.Load(p); } catch { continue; }
                foreach (var t in list) yield return t;
                continue;
            }

            if (File.Exists(p)) yield return new Track(p);
        }
    }
}

public static class PlaylistIO
{
    public static List<Track> Load(string path)
    {
        var full = Path.GetFullPath(path);
        var dir = Path.GetDirectoryName(full)!;
        var lines = ReadLines(full);
        return Path.GetExtension(full).Equals(".pls", StringComparison.OrdinalIgnoreCase)
            ? LoadPls(lines, dir)
            : LoadM3u(lines, dir);
    }

    static string[] ReadLines(string path)
    {
        var bytes = File.ReadAllBytes(path);
        string text;
        try { text = new UTF8Encoding(false, true).GetString(bytes); }
        catch (DecoderFallbackException) { text = Encoding.Latin1.GetString(bytes); }
        return text.TrimStart('﻿').Split(new[] { "\r\n", "\n", "\r" }, StringSplitOptions.None);
    }

    static List<Track> LoadM3u(string[] lines, string dir)
    {
        var list = new List<Track>();
        string? title = null;
        TimeSpan? dur = null;
        foreach (var raw in lines)
        {
            var line = raw.Trim();
            if (line.Length == 0) continue;
            if (line.StartsWith("#EXTINF:", StringComparison.OrdinalIgnoreCase))
            {
                var body = line[8..];
                int comma = body.IndexOf(',');
                var secs = (comma >= 0 ? body[..comma] : body).Trim().Split(' ')[0];
                if (double.TryParse(secs, NumberStyles.Float, CultureInfo.InvariantCulture, out var s) && s > 0)
                    dur = TimeSpan.FromSeconds(s);
                title = comma >= 0 ? body[(comma + 1)..].Trim() : null;
                continue;
            }
            if (line.StartsWith('#')) continue;
            list.Add(MakeTrack(line, dir, title, dur));
            title = null;
            dur = null;
        }
        return list;
    }

    static List<Track> LoadPls(string[] lines, string dir)
    {
        var files = new SortedDictionary<int, string>();
        var titles = new Dictionary<int, string>();
        var lengths = new Dictionary<int, double>();
        foreach (var raw in lines)
        {
            int eq = raw.IndexOf('=');
            if (eq < 0) continue;
            var key = raw[..eq].Trim();
            var val = raw[(eq + 1)..].Trim();
            if (TryKey(key, "File", out int n)) files[n] = val;
            else if (TryKey(key, "Title", out n)) titles[n] = val;
            else if (TryKey(key, "Length", out n) && double.TryParse(val, NumberStyles.Float, CultureInfo.InvariantCulture, out var l)) lengths[n] = l;
        }
        return files.Select(kv => MakeTrack(kv.Value, dir,
            titles.GetValueOrDefault(kv.Key),
            lengths.TryGetValue(kv.Key, out var l) && l > 0 ? TimeSpan.FromSeconds(l) : null)).ToList();
    }

    static bool TryKey(string key, string prefix, out int n)
    {
        n = 0;
        return key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) && int.TryParse(key[prefix.Length..], out n);
    }

    static Track MakeTrack(string entry, string dir, string? title, TimeSpan? dur)
    {
        string p = entry;
        if (!MediaFiles.IsUrl(p))
        {
            try
            {
                if (p.StartsWith("file:", StringComparison.OrdinalIgnoreCase)) p = new Uri(p).LocalPath;
                else if (!Path.IsPathRooted(p)) p = Path.GetFullPath(Path.Combine(dir, p));
            }
            catch { }
        }
        var t = new Track(p) { Duration = dur };
        if (!string.IsNullOrWhiteSpace(title)) t.Title = title;
        return t;
    }

    public static void Save(string path, IEnumerable<Track> tracks)
    {
        var sb = new StringBuilder();
        var list = tracks.ToList();
        static string Secs(Track t) => t.Duration is { } d ? ((int)d.TotalSeconds).ToString(CultureInfo.InvariantCulture) : "-1";

        if (Path.GetExtension(path).Equals(".pls", StringComparison.OrdinalIgnoreCase))
        {
            sb.AppendLine("[playlist]");
            for (int i = 0; i < list.Count; i++)
            {
                sb.AppendLine($"File{i + 1}={list[i].Path}");
                sb.AppendLine($"Title{i + 1}={list[i].DisplayTitle}");
                sb.AppendLine($"Length{i + 1}={Secs(list[i])}");
            }
            sb.AppendLine($"NumberOfEntries={list.Count}");
            sb.AppendLine("Version=2");
        }
        else
        {
            sb.AppendLine("#EXTM3U");
            foreach (var t in list)
            {
                sb.AppendLine($"#EXTINF:{Secs(t)},{t.DisplayTitle}");
                sb.AppendLine(t.Path);
            }
        }
        File.WriteAllText(path, sb.ToString(), new UTF8Encoding(false));
    }
}
